src/storage/src/hummock/compactor/iterator.rs

```diff
@@ -217,14 +263,47 @@ impl SstableStreamIterator {
             self.prune_from_valid_block_iter().await?;
         }
 
+        if !self.is_valid() {
+            return Ok(());
+        }
+
+        // Check if we need to skip the block.
+        let key = self
+            .block_iter
+            .as_ref()
+            .unwrap_or_else(|| panic!("no block iter sstinfo={}", self.sst_debug_info()))
+            .key();
+
+        if self.exceed_key_range_right(key) {
+            self.block_iter = None;
+        }
+
         Ok(())
     }
 
     pub fn key(&self) -> FullKey<&[u8]> {
-        self.block_iter
+        let key = self
+            .block_iter
             .as_ref()
             .unwrap_or_else(|| panic!("no block iter sstinfo={}", self.sst_debug_info()))
-            .key()
+            .key();
+
+        assert!(
+            !self.exceed_key_range_left(key),
+            "key {:?} key_range_left {:?}",
+            key,
+            self.key_range_left.to_ref()
+        );
+
+        assert!(
+            !self.exceed_key_range_right(key),
+            "key {:?} key_range_right {:?} key_range_right_exclusive {}",
+            key,
+            self.key_range_right.to_ref(),
+            self.key_range_right_exclusive
+        );
```

Comment:
Should this be debug_assert?

Reply:
For performance.
I think so.

For data correctness.
I think it has to be assert, when the key doesn't satisfy this condition we may give wrong output, and it can't be assumed that the user must have used FullKeyTracker externally to judge the output.
what do you think @hzxa21